### ISSUE TITLE
Container registry fails with duplicate key (HTTP 500)

### ISSUE THREAD

**Issue by Chief-Detektor** [go-gitea/gitea#27640]
### Description

Hi!
I cannot push to the integrated docker registry anymore.
It's basically #21822 but I could not reopen that issue.

I use postgres as database and tried to fix the issue with the workaround provided in the ticket above. Without any success.
~~To be honest, I need to have a working solution and I will switch to something else. I just want to let you know that this issue still exists.~~
If I can help you by providing the database dump, I can do that.

### Gitea Version

1.20.5

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

https://gist.github.com/Chief-Detektor/7b26fc63faa5bd84b5ce942fe51b6d74

### Git Version

2.40.1

### Operating System

linux

### How are you running Gitea?

I use the official helm chart.

### Database

PostgreSQL

**Comment by Chief-Detektor** [comment]
Ok I did some research:

The error occurs also on sqlite so the database is not the problem.
I tested all versions down to 1.19.4 by checking the tags out and build them locally.
Every version would give a `....manifests/latest: 404 Not Found` error which indicated to me that the issue is not directly related to gitea's registry implementation!

Out of curiosity I built and push the same Docker image from an other machine running docker 20.10.21 and that works.

Sooo I conclude that the source of my error is actually the way my docker client on macos (version 24.0.6) is pushing the images.
This is good and bad news at the same time, since it seems that only newer clients are affected and I don't know if it is even possible to determine the client's version during the push.

I have some time to investigate further and I will post my findings here.

**Comment by KN4CK3R** [comment]
It may also be helpfull if you could upload the log of all requests done by the docker client until the error happens.

**Comment by Chief-Detektor** [comment]
I found a way to reproduce the issue:
Setting the `--provenance=false` flag during image build resolves the issue. I can then push images without any issue.
So that narrows the problem down to that additional data that is produced/pushed to the registry.
Here is the log: [without the provenance flag](https://gist.github.com/Chief-Detektor/8edf0860b38d341681a588b85d988117) and after with [provenance disabled](https://gist.github.com/Chief-Detektor/d61aa746e0eea1e9776433271b23d885).

**Comment by Chief-Detektor** [comment]
I made some debugging on version v1.20.5 and the error occurs [here](https://github.com/go-gitea/gitea/blob/4126aad4aa12f3dbd4a21063d266d096ee6bb52f/routers/api/packages/container/container.go#L561).
The internal error code is "MANIFEST_BLOB_UNKOWN"

**Comment by Chief-Detektor** [comment]
When using podman as client the image can be built and pushed without any errors.
```
docker build -t <domain>/chiefdetektor/nginxtest . --push
```
as well as
```
docker build -t <domain>/chiefdetektor/nginxtest . --push --provenance=false
```
work.

The only difference is there are more entries in the package view on the admin site:
![Bildschirmfoto 2023-10-18 um 12 34 00](https://github.com/go-gitea/gitea/assets/10716635/623f3273-4d57-4c23-ae80-d0b892769718)

Detail view of the latest version:
![Bildschirmfoto 2023-10-18 um 12 35 53](https://github.com/go-gitea/gitea/assets/10716635/ba30b4ed-5ce1-44de-9341-4c8296dc8fd5)

**Comment by phaynes** [comment]
Well I am glad I finally found this post - I was about to abandon gitea and just use distribution and raw git.

While provenance is important in my use case, it would be good in the short term if the container documentation includes the line --push --provenance=false  so it is clear. 

It took ages to resolve whether the fault was in my HTTPS proxy, compose setup, gitea setup or file permissions issue.
When I did the smoke test against distribution, the push worked straight out of the box.
